### DISCUSSION TITLE
How best to handle setting build and run environment variables

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#13926] [Ideas]
While working on #13924, I had some thoughts I wanted to share.

### Description

Previously, we used the following API to control environment variables in Spack:
```python
setup_environment(self, build_env, run_env)
setup_dependent_environment(self, build_env, run_env, depedent_spec)
```
This had a lot of problems, namely that certain files/directories are only present after installation, so we can't run the same function both before and after installation. Now, we use the following API:
```python
setup_build_environment(self, env)
setup_run_environment(self, env)
setup_dependent_build_environment(self, env, dependent_spec)
setup_dependent_run_environment(self, env, dependent_spec)
```
This fixes the previously mentioned problems, but also introduces a lot of redundancy that I'll mention below.

### Proposal

I'm starting to think that a better API would be something like:
```python
setup_build_environment(self, env)
setup_run_environment(self, env)
```

Specifically, I think we should replace `setup_dependent_build_environment` with an additional call to `setup_run_environment`, and I think `setup_dependent_run_environment` isn't necessary.

### Rationale

1. `setup_run_environment` and `setup_dependent_build_environment` are almost identical in most packages that use them. This makes sense, as any env var needed to run a package is also needed when that package is being used as a build dependency. So why don't we get rid of `setup_dependent_build_environment` and just run `setup_run_environment` at build time for packages that depend on a package?
2. `setup_dependent_run_environment` should not actually be necessary. This gets into the discussion we had in #8639, but basically, if I declare something to be a `type='run'` dependency, it should be implied that any env vars needed to run a package are also needed to run a package that depends on this package. We should either autoload run-time dependencies or combine their env vars in the module file.

I'm sure there are many use cases I'm ignoring, particularly with things like `PYTHONPATH` and `PERL5LIB` where we need to recursively set env vars for language dependencies, but I think we currently overuse `setup_dependent_*_environment`. I can't think of a case where it should be used aside from an `extendable = True` package.

@alalazo @becker33 @scheibelp might be interested in this discussion.

**Comment by adamjstewart** [comment]
> I'm sure there are many use cases I'm ignoring, particularly with things like `PYTHONPATH` and `PERL5LIB` where we need to recursively set env vars for language dependencies, but I think we currently overuse `setup_dependent_*_environment`. I can't think of a case where it should be used aside from an `extendable = True` package.

To elaborate further on this, I think there are really only 2 types of environment variables, ones needed to _build_ a package and ones needed to _run_ a package. In my mind, the only point of the `setup_dependent_*_environment` functions was so that we didn't have to write:
```python
def setup_run_environment(self, env):
    env.prepend_path('PYTHONPATH', site_packages_dir)
```
in every single Python package. But if we subsume this into the `PythonPackage` base class, this is no longer an issue. So basically, I see why `setup_dependent_*_environment` can be convenient, but I don't think it's necessary.

**Comment by adamjstewart** [comment]
@alalazo @scheibelp @becker33 @tgamblin I know I opened this issue at an inopportune time (during Thanksgiving break), but I was wondering if any of you have any opinions on this. Basically what I'm proposing is that:

1. `setup_dependent_build_environment` can be replaced by a call to `setup_run_environment` for all _build_ dependencies
2. `setup_dependent_run_environment` can be replaced by a call to `setup_run_environment` for all _run_ dependencies, or gotten rid of by autoloading run dependencies via modules

See #14191 and #14198 for another unfortunate victim of this confusion.

**Comment by sethrj** [comment]
@adamjstewart Agreed, I was just refactoring a small package (not in the main spack repo) to add these methods and was struggling to discern the difference between the run environment and the dependent environments. Fortunately it's not too hard to reuse these functions:
```python
    def setup_run_environment(self, env):
        env.set('DATAPATH', self.prefix)

    def setup_dependent_build_environment(self, env, dependent_spec):
        self.setup_run_environment(env)

    setup_dependent_run_environment = setup_dependent_build_environment
```

**Comment by haampie** [comment]
Note that this distinction between `dependent_run_environment` and `run_environment` makes it unclear to me what `spack load x` and `spack env activate x` is supposed to do... run both for all transitive run type deps?

I would expect `setup_run_environment` to be executed for all transitive run type deps, as well as `setup_dependent_run_environment`, but this was not the default behavior for a long time. In https://github.com/spack/spack/issues/23368 it was confirmed that this is the expected behavior.

Then what's the point of setup_dependent_run_environment except for causing confusion?

**Comment by adamjstewart** [comment]
Another question is how to handle tests, both at build-time (`spack install --test=root`) and at run-time (`spack test run`). Personally I would like these two phases to run `setup_run_environment`.

**Reply by haampie** [reply]
I would expect them to, they don't?

**Reply by adamjstewart** [reply]
I wish. `spack install --test=root` uses `setup_build_environment` and `spack test run` uses `setup_run_environment` (despite both tests running the same thing. In most package where I need to set any environment variable (such as [py-pythran](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-pythran/package.py)) I end up setting it 4 times.

**Reply by tldahlgren** [reply]
I believe they do. A problem with test is we aren't saving (or keeping) enough info in the database.
